### DISCUSSION TITLE
Add support for snapcast

### DISCUSSION THREAD

**Discussion by Schluggi** [music-assistant/support#381] [Feature requests and ideas]
Hi,

I'm using snapcast for a while and i would love to see this as an input source for Music Assistant. 

This also would add a lot of more input capabilities because of the possibilities of snapcast.

https://github.com/badaix/snapcast

JFYI: I also created an [issue](https://github.com/badaix/snapcast/issues/1019) on snapcast side to add  "play from url" functionality to use snapcast also as a speaker device.

**Comment by marcelveldt** [comment]
For snapcast we can even do better because Music Assistant internally works at the PCM level. We could forward those PCM frames directly to snapcast bypassing the intermediate HTTP transport

**Reply by foxy82** [reply]
@marcelveldt  nice work on Music Assistant.

If music assistant is serious about being a multiroom/grouped audio experience with everything in sync wouldn't it have a better chance of being in sync if you bypassed the snapcast server and sent directly to the snapcast clients? That was music assistant could also manage which snapcast client sources are sent which streams?

**Reply by marcelveldt** [reply]
I'm not going to reinvent any wheels and rebuild the snapcast server. From what I see is that the snapcast server can be controlled to also control which streams go where.

**Comment by koying** [comment]
Was about to ask the same ;)

~~Actually, Snapcast supports FLAC as well, so it would seem to be relatively easy to just "push" the chunks to snapcast via its TCP interface in addition to creating a "pull" web server for the traditional servers.~~

Nevermind, the TCP interface only takes PCM.

**Reply by marcelveldt** [reply]
Yeah, that was the direction I was thinking. If somebody can add that to the HA integration of Snapcast we're all set ;-)

**Reply by koying** [reply]
What would need to be added to the Snapcast integration, exactly?
Or do you mean the integration acting as proxy?

**Reply by marcelveldt** [reply]
The play_media service would need to be implemented for regular "play url" calls.
In that case it will be able to support Music Assistant and to play any other media from HA to snapcast clients.
If Snapcast supports the "play_media" service where a url can be sent no changes are needed on the MA side.

That said, it looks like to properly support Airplay I do need to make some changes to the MA code to support sending the PCM (or encoded audio) chunks somewhere directly (bypassing the webserver) so maybe we should go that route for snapcast too and skip the intermediate webserver.

I can make all changes on the Music Assistant side but it would be great if someone else wants to pickup the changes needed in the Snapcast integration. The MA integration has direct access to the HA entity instance and bypasses the state machine. See "player_controls.py" which contains a few more player specific adjustments.

I am thinking about a callback to deliver audio chunks (pcm or encoded to a format of choice)

**Reply by marcelveldt** [reply]
So yes, to answer your question: The HA integration for Snapcast would be acting as a proxy. 
To support playback to snapcast targets MA would become a requirement but I don't think that is an issue.

**Reply by foxy82** [reply]
> That said, it looks like to properly support Airplay I do need to make some changes to the MA code to support sending the PCM (or encoded audio) chunks somewhere directly (bypassing the webserver) so maybe we should go that route for snapcast too and skip the intermediate webserver.

Does this mean you will need a service like `play_pcm` to distingish it fom the `play_url`

**Reply by marcelveldt** [reply]
Yes, that is a possible direction.

**Comment by jmound** [comment]
@marcelveldt For snapcast support- what do you think of setting up gstreamer to play by url, and having gstreamer send the audio to a listening snapserver?

If so, what is the url that gstreamer (playbin?) should be pointing to?

**Reply by marcelveldt** [reply]
That is not needed at all because Music Assistant can feed raw pcm audio directly to the snapcast server. 

Another idea is to just embed the snap server into music assistant and combine both worlds.

**Reply by jmound** [reply]
I thought that wouldn't work without a "play from url" function that snapcast doesn't have? I think embedding the snapserver in MA makes sense if it will work- with the added benefit of getting multi player sync "for free". 

Regarding "can feed raw pcm audio directly to the snapcast server" - is that possible now? If so, how would a snapserver access that?

**Comment by petanque** [comment]
I don't know if this should have it's own discussion, but what about adding a snapcast client so it's possible to stream audio from a snapcast server?

**Reply by OzGav** [reply]
Why would you want this versus having MA do the music serving?

**Reply by petanque** [reply]
Wife factor, she is used to apps like spotify, youtube and Audiobookshelf as it is easier for her.
I use both the snapcast client and the server on all my satellites so that they prioritize which sound is to be played and she therefore only has to connect and does not need to change the input or mess with other settings.
But it would clearly be better if it were possible to do everything in MA and not on the satellites.

**Reply by OzGav** [reply]
The MA UI is a work in progress but improving all the time so perhaps you could show her that. Can't help with audiobookshelf yet though...

**Reply by petanque** [reply]
I really like MA and look forward to seeing it develop, it has made me ditch the Logitech Media Server I was using before.
Wives are not easy to please because she couldn't like that either, but she thinks MA ui looks better so there is hope.

**Comment by palitu** [comment]
really cool to see this!  I too use snapcast, and it has been solid for years

As an aside - for those snapcasters, what are you using for client hardware?  i am pretty much rPi0/3 with amp hats plugging into speakers, but woulld love something a little more professional looking and secure

ta

**Reply by SantiagoSotoC** [reply]
I am using android devices, from google tv to some old tablets. 

Also snapcast client is being ported to esp32 without psram with support for flac streams, I haven't tried it yet. I have an order of esps and some das i2s but it's going to take a month. 

[https://github.com/pschatzmann/arduino-snapclient](https://github.com/pschatzmann/arduino-snapclient)

If I have good results I could create a specific board, but for now I can't promise it.

**Reply by 77Silvermane77** [reply]
I'm new to HA, just trying it out on a temp rig until my new N100 based toy gets delivered today. Just discovered MA and will be trying that out too as it sounds very promising! I'm migrating away from Hubitat party because of the dashboards but also because I wanted to run Z2MQTT and Snapcast. If MA integrates multiroom audio, either to Snapcast- and/or MA-clients, that would be amazing.

At the risk of going further off topic: @SantiagoSotoC, on the ESP32 front that's great that they're porting snapcast, but all my ESP32 devices run WLED. I'll read up on that but just off the top of my head it would be insane if WLED and snapcast could be combined so that WLED could do music animations directly from the snapcast stream, rather than having to rely on a Mic or other audio input :)

**Reply by OzGav** [reply]
There is someone trying to add Snapcast support at the moment. No idea on the timeframe yet though…

**Reply by SantiagoSotoC** [reply]
I`m integrating snapcast into music assiatant. There is already a merge version to main. In the next beta I guess it will come out, but that depends on @marcelveldt .

Regarding wled with snapcast in the same esp32 I think it's a bit difficult. For the cost I think it is better to use two esp32 and split the functionality. Both are demanding projects for the esp32

**Reply by DerPicknicker** [reply]
@SantiagoSotoC ... 

How should the SNAPCAST config look to use with MA? 

Best regards

**Reply by OzGav** [reply]
Sorry we don’t understand the question

**Reply by SantiagoSotoC** [reply]
@OzGav  This has already been discussed in the new discussion.
@DerPicknicker The default configuration should work just make sure you use version 0.27.0.

**Comment by LostName1** [comment]
This addition would be so nice, I hope this get added as it would improve the actual functionalities of Music Assistant.

Great idea from the dev.

**Comment by OzGav** [comment]
Snapcast support will be added in 2.0.0.b76. Docs here https://github.com/orgs/music-assistant/discussions/1668
